runtime/dart_isolate.cc

```diff
@@ -332,6 +334,34 @@ fml::RefPtr<fml::TaskRunner> DartIsolate::GetMessageHandlingTaskRunner() const {
   return message_handling_task_runner_;
 }
 
+bool DartIsolate::LoadLoadingUnit(
+    intptr_t loading_unit_id,
+    std::unique_ptr<const fml::SymbolMapping> snapshot_data,
+    std::unique_ptr<const fml::SymbolMapping> snapshot_instructions) {
```

Comment:
A subtle error has been introduced here when you went from the C++ API to the Dart C API. The symbol mapping owns the library handle you created later in this patch. However, both go out of scope here. The library is now un-reffed and may close the library handle. This can cause issues depending on the dynamic linker implementation. From the docs of `dlcose`:

```
"The use of dlclose() reflects a statement of intent on the part of the process,
but does not create any requirement upon the implementation, such as removal
of the code or symbols referenced by handle. Once an object has been closed
using dlclose() an application should assume that its symbols are no longer
available to dlsym()."
```

Depending on the implementation, subject to timing issues within the implementation, and, other references to this library by the engine, the OS is also free to unmap the memory you just gave the Dart code via the C API which would then lead to segfaults.

We do this in other places by owning a reference to a `flutter::DartSnapshot` which is nothing but two mappings. I suggest you do the same. Maybe toss it inside a `set` in `DartIsolate`.

I am glad we are running into these design issues now. Using the C++ wrappers forces you to think about lifecycle issues that raw pointers don't. This would be a very hard to debug segfault otherwise.

Reply:
Ahh, I see what you meant by lifecycle now. Thanks for catching this and the explanation.

Reply:
Right. When you vend a raw pointer, you need to know upfront till when the user expects to use the pointer. That way, you can keep what is being pointed to alive for at least that long.

From my last comment on lifecycle, it is still not clear to me if the VM can use this post isolate termination. Presumably not, but just good to check.